File: topoflow/actions.py
# SPDX-License-Identifier: AGPL-3.0-only
# Copyright (C) 2026 Discover Legal
"""Legal-action enumeration A(s) (spec §5) [AF] + [NEW].

Edges are never arms: the topology family is a handful of (mode, tau, round)
arms, not per-edge choices. With 2 modes, 4 tau buckets, 3 round buckets the
topology family is 1 + 4*3 = 13 arms; plus ~14 invoke variants + a few skips
≈ ~30 arms total.
"""
from __future__ import annotations

from dataclasses import dataclass, field

from .config import Config
from .types import Action


@dataclass
class Plan:
    """Tracks which optional aux cells are still schedulable."""

    optional: list[str] = field(default_factory=list)
    invoked: set[str] = field(default_factory=set)
    skipped: set[str] = field(default_factory=set)

    def still_scheduled(self) -> list[str]:
        return [c for c in self.optional if c not in self.invoked and c not in self.skipped]

    def mark_invoked(self, cell: str) -> None:
        self.invoked.add(cell)

    def mark_skipped(self, cell: str) -> None:
        self.skipped.add(cell)


def cell_enabled(skill: str, plan: Plan, cfg: Config) -> bool:
    # optional aux cells may be invoked at most once
    if skill in cfg.optional_aux:
        return skill not in plan.invoked and skill not in plan.skipped
    # solvers and the evaluator are always invocable (budget bounds them)
    return True


def legal_actions(sig, handoff, plan: Plan, budget, cfg: Config) -> list[Action]:
    A: list[Action] = []
    # [AF] invoke(skill, model)
    for k, m in cfg.variant_pool:
        if cell_enabled(k, plan, cfg):
            A.append(Action("invoke", skill=k, model=m))
    # [NEW] topology(mode, tau_bucket, k_in, round_bucket)
    for mode in cfg.topo_modes:
        if mode == "linear":
            A.append(Action("topology", topo_mode="linear"))
        elif mode == "dytopo":
            for tb in range(len(cfg.tau_buckets)):
                for rb in range(len(cfg.round_buckets)):
                    A.append(
                        Action(
                            "topology",
                            topo_mode="dytopo",
                            tau_bucket=tb,
                            k_in=cfg.dytopo_k_in,
                            round_bucket=rb,
                        )
                    )
    # [AF] skip:X for still-scheduled optional cells, only if >=1 other legal action remains
    if A:
        for X in plan.still_scheduled():
            A.append(Action("skip", target=X))
    # [AF] terminate is never emitted here — it is implicit (§9)
    return prune_to_keep_finishable(A, plan, budget, cfg)


def prune_to_keep_finishable(A: list[Action], plan: Plan, budget, cfg: Config) -> list[Action]:
    """Guarantee a termination path exists. The evaluator invoke (always enabled)
    provides it; if for any reason it is absent, add it."""
    has_eval = any(a.kind == "invoke" and a.skill == "evaluator" for a in A)
    if not has_eval:
        A.append(Action("invoke", skill="evaluator", model=cfg.default_model_for("evaluator")))
    return A

File: biglaw-go/internal/topoflow/actions.go
// SPDX-License-Identifier: AGPL-3.0-only
// Copyright (C) 2026 Discover Legal

package topoflow

// Plan tracks which optional aux cells are still schedulable.
type Plan struct {
	optional []string
	invoked  map[string]bool
	skipped  map[string]bool
}

// NewPlan builds a plan over the optional aux cells.
func NewPlan(optional []string) *Plan {
	return &Plan{optional: append([]string(nil), optional...), invoked: map[string]bool{}, skipped: map[string]bool{}}
}

func (p *Plan) stillScheduled() []string {
	var out []string
	for _, c := range p.optional {
		if !p.invoked[c] && !p.skipped[c] {
			out = append(out, c)
		}
	}
	return out
}

func (p *Plan) markInvoked(cell string) { p.invoked[cell] = true }
func (p *Plan) markSkipped(cell string) { p.skipped[cell] = true }

func cellEnabled(skill string, plan *Plan, cfg Config) bool {
	if cfg.IsOptionalAux(skill) {
		return !plan.invoked[skill] && !plan.skipped[skill]
	}
	return true // solvers + evaluator always invocable (budget bounds them)
}

// legalActions enumerates A(s) (spec §5). Edges are never arms.
func legalActions(plan *Plan, cfg Config) []Action {
	var A []Action
	for _, km := range cfg.VariantPool() {
		if cellEnabled(km[0], plan, cfg) {
			A = append(A, InvokeAction(km[0], km[1]))
		}
	}
	for _, mode := range cfg.TopoModes {
		switch mode {
		case "linear":
			A = append(A, LinearTopoAction())
		case "dytopo":
			for tb := range cfg.TauBuckets {
				for rb := range cfg.RoundBuckets {
					A = append(A, DytopoAction(tb, cfg.DytopoKIn, rb))
				}
			}
		}
	}
	// skip:X only if >=1 other legal action remains
	if len(A) > 0 {
		for _, X := range plan.stillScheduled() {
			A = append(A, SkipAction(X))
		}
	}
	return pruneToKeepFinishable(A, cfg)
}

// pruneToKeepFinishable guarantees a termination path (the evaluator invoke).
func pruneToKeepFinishable(A []Action, cfg Config) []Action {
	for _, a := range A {
		if a.Kind == "invoke" && a.Skill == "evaluator" {
			return A
		}
	}
	return append(A, InvokeAction("evaluator", cfg.DefaultModelFor("evaluator")))
}
